Handle endpoints with the same name

rjschwei opened the pull request:
AWS has introduced duplication in endpoint names in the metadata such as iam/info and identity-credentials/ec2/info. Previously the first endpoint found would get added to the CLI options and the second one would get dropped. The new code will use the path to differentiate the query options and collect all the data when no query option is specified.

rjschwei committed:
Handle endpoints with the same name

AWS has introduced duplication in endpoint names in the metadata such as
iam/info and identity-credentials/ec2/info. Previously the first endpoint
found would get added to the CLI options and the second one would get dropped.
The new code will use the path to differentiate the query options and collect
all the data when no query option is specified.

bear454 reviewed (approved):
Thanks for the quick fix.

smarlowucf commented on lib/ec2metadata/__init__.py:
```diff
@@ -68,11 +68,28 @@ def _add_mata_option(self, path):
-                if item not in options:
-                    if item[-1] != '/':
+                if item[-1] == '/':
+                    self._add_mata_option(path+item)
```
Pre-existing typo: s/mata/meta

smarlowucf commented on lib/ec2metadata/__init__.py:
```diff
@@ -68,11 +68,28 @@ def _add_mata_option(self, path):
+                        cnt = 1
+                        while option_name in options:
+                            option_name = '-'.join(
+                                prefix_name_opts[:cnt] + [option_name]
```
Not sure I follow logic without context but it seems (could be wrong) like we end up with two options:

- info
- ec2-identity-credentials-info

If there's any clash I would expect both names become verbose and I wonder if the dash in identitycredentials causes any trouble?

Expectation would be:

- iam-info
- ec2-identity-credentials-info

Do we split this back on dash? Then identity-credentials/ec2/info becomes identity/credentials/ec2/info?

rjschwei replied:
Well there is no way to predict if there is a clash. When the code process the first endpoint called `info` it does not know that there will be another.

If we expect the expansion of both names, which is a good suggesting that can be added but requires a bit more refactoring as we wold then have to remove the previously generated option name and insert the newly expanded name. But then if a third one comes along we are back to the short name, i.e. `info`. That can only be accounted for with an additional tracking list `duplicate_names` which in and off itself if not a problem.

But generally this is a good idea as the current approach of "first one wins the naming battle" is flaky in the sense that if IMDS changes the order in which things are presented than `--info` means one thing today and may mean a different thing tomorrow.

I'll do some refinement

rjschwei committed:
- Update version
- Fix typo

rjschwei committed:
Eliminate ambigutity

When an endpoint exists expand the option name of the existing end point and
the endpoint found with the same name. Remove the existing entry. This avoids
providing the same information with 2 options and eliminates order dependency
on the metadata information as the first entry found would have occupied the
conflicting name which may have triggered flip-flop.

bear454 reviewed (commented):
If I read correctly, `--info` will no longer be a valid argument, replaced by:

* `--iam-info`
* `--ec2-info`

Is that a correct interpretation?

rjschwei commented:
> If I read correctly, `--info` will no longer be a valid argument, replaced by:
> 
> * `--iam-info`
> * `--ec2-info`
> 
> Is that a correct interpretation?

Yes that is correct. There is a pending PR for suse-rancher-setup to deal with this incompatible change.
